### PR TITLE
Add mypy_types benchmark

### LINKED ISSUE #35: Add a benchmark that heavily uses type annotations
This is a feature request.

Would it make sense to add a benchmark for code that _heavily_ uses annotations, generics, and other ``typing`` features? I even have a suggestion for such benchmark: running ``mypy`` on itself (on its own code). The run currently takes around 10-20 sec. (depending on machine). If this is too long, other option would be to run mypy on one of typeshed stubs, this will take less than a second.

### PR DESCRIPTION
**Pull request by Fidget-Spinner**
Fixes #35. Ported from Pyston.

### CODE REVIEW

**Comment by Fidget-Spinner** [comment]
Huh, I forgot that mypy doesn't support pypy due to dependence on CPython AST. I wonder if there's a way to skip installation of mypy and this specific benchmark just for pypy?

**Review by vstinner — COMMENTED** [review]
Can you try not not add the benchmark to DEFAULT_GROUP if you are running PyPy?

**Comment by Fidget-Spinner** [comment]
Yay the tests are finally passing! This should ready for review now. Please take a look when you have the time Victor.

File: pyperformance/requirements.in [hunk 1]
```diff
@@ -32,3 +32,4 @@ tornado                            # bm_tornado_http
 
 psutil
 dulwich                            # bm_dulwich_log
+mypy                               # bm_mypy_types
```
Comment (gvanrossum) [inline pyperformance/requirements.in:35]: Add that newline here.

**Review by gvanrossum — COMMENTED** [review]
LGTM. @ericsnowcurrently take note!

**Comment by gvanrossum** [comment]
I just was reminded of this PR again. Do we still need to do this when we have Eric's more general #109?

> Recently I remembered mypy uses mypyc for speedups. I'm inclined to pass `--no-binary` to pip so that the benchmark runs more python code (and less C extension code). WDYT?

Oh, definitely. Benchmarking mypy with mypyc's speedups says more about mypyc than about the Python interpreter used to run it.

> Though I'm not sure how to pass in that argument just for mypy. Seems like `pip-compile` will add the argument to all other packages, which will affect django among others.

Maybe #109 will help there too? It has a separate toml file per benchmark. I believe Eric also has prepared mods for the Pyston benchmarks that makes them runnable with his new infrastructure; I think it would be a shame if the Pyston benchmarks were forked. We should probably talk to the Pyston team about this before proceeding (@kmod, @undingen).

File: pyperformance/requirements.txt [hunk 2]
```diff
@@ -24,6 +24,10 @@ markupsafe==1.1.1
     # via mako
 mpmath==1.2.1
     # via sympy
+mypy-extensions==0.4.3
+    # via mypy
+mypy==0.812
```
Comment (pablogsal) [inline pyperformance/requirements.txt:29]: This is probably going to pick up the extension if the wheel is available and the pure python if not, which may give different numbers in different versions. Am I mistaken? Maybe I am missing something.

Also, is your intent to test the extension?

**Comment by vstinner** [comment]
> This is probably going to pick up the extension if the wheel is available and the pure python if not, which may give different numbers in different versions. Am I mistaken? Maybe I am missing something.

If you really don't want to measure the C implementation of mypy, the benchmark should have an option to decide which implementations is taken. See for example pyperformance/benchmarks/bm_pickle.py which implements such option and raises an error if it gets the wrong implementation.

**Comment by kmod** [comment]
fwiw in Pyston we run the benchmark without mypyc, but get that not 100% by choice because only the prebuilt mypy wheels are built with mypyc and building a fresh wheel yourself (like we need to do) will not use it.

**Comment by ericsnowcurrently** [comment]
FYI, I've merged my PR that allows running benchmarks that aren't part of pyperformance.  I also have a PR up against the Pyston benchmarks repo to allow them to be run using pyperformance: https://github.com/pyston/python-macrobenchmarks/pull/3.

**Comment by Fidget-Spinner** [comment]
@ericsnowcurrently awesome! I'll close this PR then since Pyston has the same.

**Pull request closed without merging**